File: internal/tools/weather.go
package tools

import (
	"encoding/json"
	"fmt"

	"github.com/go-resty/resty/v2"
)

type WeatherTool struct{}

func NewWeatherTool() ToolsFactory {
	return &WeatherTool{}
}

type WeatherArguments struct {
	Location string `json:"location"`
	Unit     string `json:"unit"`
}

func (w *WeatherTool) CallTool(arguments string) string {
	var args WeatherArguments
	if err := json.Unmarshal([]byte(arguments), &args); err != nil {
		return fmt.Sprintf("Error parsing arguments: %v", err)
	}

	baseURL := fmt.Sprintf("http://wttr.in/%s?format=j1", args.Location)

	client := resty.New()

	var result map[string]interface{}
	resp, err := client.R().SetResult(&result).Get(baseURL)
	if err != nil {
		return fmt.Sprintf("Error making request: %v", err)
	}

	if resp.StatusCode() != 200 {
		return fmt.Sprintf("Error response from API: %s", resp.Status())
	}

	currentCondition, ok := result["current_condition"].([]interface{})
	if !ok || len(currentCondition) == 0 {
		return "No current weather data available."
	}

	var temp string
	var tempFeelsLike string
	var weatherDesc string
	currentWeather := currentCondition[0].(map[string]interface{})
	if args.Unit == "celsius" {
		temp = currentWeather["temp_C"].(string)
		tempFeelsLike = currentWeather["FeelsLikeC"].(string)
	} else {
		args.Unit = "fahrenheit"
		temp = currentWeather["temp_F"].(string)
		tempFeelsLike = currentWeather["FeelsLikeF"].(string)
	}

	weather := currentWeather["weatherDesc"].([]interface{})
	for _, item := range weather {
		desc, _ := item.(map[string]interface{})
		weatherDesc += desc["value"].(string) + "\n"
	}

	return fmt.Sprintf(
		"The current weather in %s is %s and temperature is %s°%s (feels like %s°%s).",
		args.Location, weatherDesc, temp, args.Unit, tempFeelsLike, args.Unit,
	)
}

File: .teo/skills/weather/scripts/weather.py
import json
import sys
import urllib.request
import urllib.parse
import urllib.error

def main():
    if len(sys.argv) < 2:
        print("Error: No arguments provided")
        sys.exit(1)

    try:
        params = json.loads(sys.argv[1])
    except json.JSONDecodeError:
        print("Error: Invalid JSON argument")
        sys.exit(1)

    location = params.get("location")
    unit = params.get("unit", "celsius")

    if not location:
        print("Error: location is required")
        sys.exit(1)

    # Encode location for URL
    encoded_location = urllib.parse.quote(location)
    url = f"http://wttr.in/{encoded_location}?format=j1"

    try:
        with urllib.request.urlopen(url) as response:
            if response.status != 200:
                print(f"Error response from API: {response.status}")
                return
            
            data = json.loads(response.read().decode('utf-8'))
            
            current_condition = data.get("current_condition", [])
            if not current_condition:
                print("No current weather data available.")
                return

            current = current_condition[0]
            
            temp = ""
            temp_feels_like = ""
            
            if unit == "fahrenheit":
                temp = current.get("temp_F")
                temp_feels_like = current.get("FeelsLikeF")
            else:
                temp = current.get("temp_C")
                temp_feels_like = current.get("FeelsLikeC")
                
            weather_desc_list = current.get("weatherDesc", [])
            weather_desc = ""
            for item in weather_desc_list:
                weather_desc += item.get("value", "") + "\n"
            weather_desc = weather_desc.strip()

            print(f"The current weather in {location} is {weather_desc} and temperature is {temp}°{unit} (feels like {temp_feels_like}°{unit}).")

    except urllib.error.HTTPError as e:
        print(f"Error making request: {e}")
    except Exception as e:
        print(f"Error making request: {e}")

if __name__ == "__main__":
    main()
